File: src/simple/test/test_simple_solver_handle.cpp
/*
 * Copyright 2017-2018, UT-Battelle, LLC
 *
 * SPDX-License-Identifier: BSD-3-Clause
 * License-Filename: LICENSE
 */
#include <Teuchos_DefaultComm.hpp>
#include <Teuchos_XMLParameterListHelpers.hpp>
#include <Teuchos_StandardCatchMacros.hpp>
#include <Teuchos_Array.hpp>
#include <Tpetra_Map.hpp>
#include <Tpetra_CrsMatrix.hpp>
#include <Tpetra_MultiVector.hpp>

#include <iostream>

#include "solver_handle.hpp"

int main(int argc, char *argv[]) {
  bool success = false;
  bool verbose = true;

  try {
    using Teuchos::RCP;
    using Teuchos::rcp;
    using Teuchos::rcp_dynamic_cast;
    using Teuchos::ParameterList;
    using Teuchos::tuple;

    typedef double                                  SC;
    typedef int                                     LO;
    typedef long long                               GO;
    typedef Kokkos::Compat::KokkosSerialWrapperNode NO;

    using Map         = Tpetra::Map<LO,GO,NO>;
    using MultiVector = Tpetra::MultiVector<SC,LO,GO,NO>;
    using Matrix      = Tpetra::CrsMatrix<SC,LO,GO,NO>;

    using STS = Teuchos::ScalarTraits<SC>;

    // Initialize MPI system
    Teuchos::GlobalMPISession mpiSession(&argc, &argv, NULL);
    RCP<const Teuchos::Comm<int> > comm = Teuchos::DefaultComm<int>::getComm();
    int myRank   = comm->getRank();
    int numProcs = comm->getSize();

    // Read in the parameter list
    ParameterList paramList;
    Teuchos::updateParametersFromXmlFileAndBroadcast("stratimikos.xml", Teuchos::Ptr<ParameterList>(&paramList), *comm);

    // Set parameters
    const int numMyElements = 50;
    int numGlobalElements = numMyElements*comm->getSize();

    // Step 0: Construct tri-diagonal matrix
    RCP<Map>    rowMap = Teuchos::rcp(new Map(numGlobalElements, numMyElements, 0, comm));
    RCP<Matrix> A = Teuchos::rcp(new Matrix(rowMap, 3));
    for (LO lclRow = 0; lclRow < static_cast<LO>(numMyElements); ++lclRow) {
      const GO gblRow = rowMap->getGlobalElement (lclRow);

      if (gblRow == 0) {
        // A(0, 0:1) = [2, -1]
        A->insertGlobalValues (gblRow,
                               tuple<GO> (gblRow, gblRow + 1),
                               tuple<SC> (2.0, -1.0));
      } else if (static_cast<Tpetra::global_size_t> (gblRow) == numGlobalElements - 1) {
        // A(N-1, N-2:N-1) = [-1, 2]
        A->insertGlobalValues (gblRow,
                               tuple<GO> (gblRow - 1, gblRow),
                               tuple<SC> (-1.0, 2.0));
      } else {
        // A(i, i-1:i+1) = [-1, 2, -1]
        A->insertGlobalValues (gblRow,
                               tuple<GO> (gblRow - 1, gblRow, gblRow + 1),
                               tuple<SC> (-1.0, 2.0, -1.0));
      }
    }
    A->fillComplete();

    // The solution is iota
    RCP<MultiVector> rhs = Teuchos::rcp(new MultiVector(rowMap, 1));
    rhs->putScalar(0.0);
    auto data = rhs->getDataNonConst(0);
    if (myRank == 0)
      data[0] = -1.0;
    if (myRank == numProcs-1)
      data[numMyElements-1] = numGlobalElements;

    RCP<MultiVector> lhs       = Teuchos::rcp(new MultiVector(rowMap, 1));
    RCP<MultiVector> lhs_exact = Teuchos::rcp(new MultiVector(rowMap, 1));
    lhs->randomize();
    data = lhs_exact->getDataNonConst(0);
    for (int i = 0; i < numMyElements; i++)
      data[i] = myRank*numMyElements + i;

    // Step 1: initialize a handle
    ForTrilinos::TrilinosSolver handle;
    handle.init(comm);

    // Step 2: setup the problem
    handle.setup_matrix(A);

    // Step 3: setup the solver
    handle.setup_solver(Teuchos::rcpFromRef(paramList));

    // Step 4: solve the system
    handle.solve(rhs, lhs);

    // Check the solution
    Teuchos::Array<typename STS::magnitudeType> norms(1);
    lhs->update(-1.0, *lhs_exact, 1.0);
    lhs->norm2(norms);

    // TODO: Get the tolerance out of the parameter list
    TEUCHOS_ASSERT(norms[0] < 1e-6);

    // Step 5: clean up
    handle.finalize();

    success = true;
  }
  TEUCHOS_STANDARD_CATCH_STATEMENTS(verbose, std::cerr, success);

  return ( success ? EXIT_SUCCESS : EXIT_FAILURE );
}

File: src/simple/test/test_simple_solver_handle.f90
! Copyright 2017-2018, UT-Battelle, LLC
!
! SPDX-License-Identifier: BSD-3-Clause
! License-Filename: LICENSE
module myoperators
  use forteuchos
  use fortpetra
  implicit none

  type, extends(ForTpetraOperator) :: TriDiagOperator
    type(TpetraMap) :: domain_map
    type(TpetraMap) :: range_map
  contains
    procedure :: apply => my_apply
    procedure :: getDomainMap => my_getDomainMap
    procedure :: getRangeMap => my_getRangeMap
  end type
  interface TriDiagOperator
    procedure new_TriDiagOperator
  end interface

contains
  function new_TriDiagOperator(domain_map, range_map) &
      result(self)
    use, intrinsic :: ISO_C_BINDING
    type(TpetraMap), intent(in) :: domain_map
    type(TpetraMap), intent(in) :: range_map
    type(TriDiagOperator) :: self
    write(*,*) "Called constructor"
    self%domain_map = domain_map
    self%range_map = range_map
  end function

  subroutine my_apply(self, x, y, mode, alpha, beta)
    use, intrinsic :: ISO_C_BINDING
    implicit none
    class(TriDiagOperator), intent(in) :: self
    class(TpetraMultiVector), intent(in) :: x
    class(TpetraMultiVector), intent(inout) :: y
    integer(kind(TeuchosETransp)), intent(in) :: mode
    real(scalar_type), intent(in) :: alpha
    real(scalar_type), intent(in) :: beta
    integer(size_type) :: n, i
    real(scalar_type), dimension(:), pointer :: xdata
    real(scalar_type), dimension(:), pointer :: ydata

    xdata => x%getData(INT(1,size_type))
    ydata => y%getData(INT(1,size_type))
    n = x%getLocalLength()

    ! y = alpha * A*x + beta * y
    do i = 1, n
      ydata(i) = beta * ydata(i)

      ! A has [-1 2 -1] stencil
      ydata(i) = ydata(i) + 2*alpha*xdata(i)
      if (i > 1) then
        ydata(i) = ydata(i) - alpha*xdata(i-1)
      end if
      if (i < n) then
        ydata(i) = ydata(i) - alpha*xdata(i+1)
      end if
    end do

  end subroutine

  function my_getDomainMap(self) &
      result(domain_map)
    use, intrinsic :: ISO_C_BINDING
    implicit none
    class(TriDiagOperator), intent(in) :: self
    type(TpetraMap) :: domain_map

    domain_map = self%domain_map
  end function

  function my_getRangeMap(self) &
      result(range_map)
    use, intrinsic :: ISO_C_BINDING
    implicit none
    class(TriDiagOperator), intent(in) :: self
    type(TpetraMap) :: range_map

    range_map = self%range_map
  end function

end module


program main

#include "ForTrilinosSimpleInterface_config.hpp"
#include "ForTrilinos.h"

  use ISO_FORTRAN_ENV
  use, intrinsic :: ISO_C_BINDING

#ifdef HAVE_MPI
  use mpi
#endif

#include "ForTrilinos.h"
  use fortrilinos
  use forteuchos
  use fortpetra
  use fortest
  use myoperators
  implicit none

  integer(int_type) :: my_rank, num_procs

  integer(global_size_type) :: n_global
  integer(size_type) :: n, max_entries_per_row, num_vecs = 1, lda
  integer(int_type) :: row_nnz

  integer :: ierr
  integer(local_ordinal_type) :: i
  integer(global_ordinal_type) :: offset
  real(scalar_type) :: one = 1.0
  real(norm_type) :: norm

  type(TeuchosComm) :: comm
  type(ParameterList) :: plist
  type(TrilinosSolver) :: solver_handle
  type(TpetraMap) :: map
  type(TpetraCrsMatrix) :: A
  type(TpetraMultiVector) :: B, X, Xtrue
  class(ForTpetraOperator), allocatable :: op

  real(scalar_type), dimension(:), allocatable :: lhs, rhs
  real(norm_type), dimension(:), allocatable :: norms
  integer(global_ordinal_type), dimension(:), allocatable :: cols
  real(scalar_type), dimension(:), allocatable :: vals

  n = 50

#ifdef HAVE_MPI
  ! Initialize MPI subsystem
  call MPI_INIT(ierr)
  if (ierr /= 0) then
    write(*,*) "MPI failed to init"
    stop 1
  endif

  comm = TeuchosComm(MPI_COMM_WORLD)
#else
  comm = TeuchosComm()
#endif

  my_rank = comm%getRank()
  num_procs = comm%getSize()

  write(*,*) "Processor ", my_rank, " of ", num_procs

  ! Read in the parameterList
  plist = ParameterList("Stratimikos"); FORTRILINOS_CHECK_IERR()
  call load_from_xml(plist, "stratimikos.xml"); FORTRILINOS_CHECK_IERR()

  ! ------------------------------------------------------------------
  ! Step 0: Construct tri-diagonal matrix
  n_global = -1
  map = TpetraMap(n_global, n, comm); FORTRILINOS_CHECK_IERR()

  max_entries_per_row = 3
  A = TpetraCrsMatrix(map, max_entries_per_row, TpetraDynamicProfile)

  allocate(cols(max_entries_per_row))
  allocate(vals(max_entries_per_row))
  offset = n * my_rank
  do i = 1, n
    row_nnz = 1
    if (i .ne. 1 .or. my_rank > 0) then
      cols(row_nnz) = offset + i-1
      vals(row_nnz) = -1.0
      row_nnz = row_nnz + 1
    end if
    cols(row_nnz) = offset + i
    vals(row_nnz) = 2.0
    row_nnz = row_nnz + 1
    if (i .ne. n .or. my_rank .ne. num_procs-1) then
      cols(row_nnz) = offset + i+1
      vals(row_nnz) = -1.0
      row_nnz = row_nnz + 1
    end if

    call A%insertGlobalValues(offset + i, cols(1:row_nnz-1), vals(1:row_nnz-1)); FORTRILINOS_CHECK_IERR()
  end do
  call A%fillComplete(); FORTRILINOS_CHECK_IERR()

  ! This automatically zeroes out X
  X = TpetraMultiVector(map, num_vecs); FORTRILINOS_CHECK_IERR()

  ! The solution X(i) = i-1
  allocate(lhs(n))
  allocate(rhs(n))
  if (my_rank > 0) then
    rhs(1) = 0.0
  else
    rhs(1) = -1.0
  end if
  if (my_rank .ne. num_procs-1) then
    rhs(n) = 0.0
  else
    rhs(n) = offset+n
  end if
  do i = 2, n-1
    rhs(i) = 0.0
  end do
  do i = 1, n
    lhs(i) = offset + i-1
  end do
  lda = n

  Xtrue = TpetraMultiVector(map, lhs, lda, num_vecs); FORTRILINOS_CHECK_IERR()
  B = TpetraMultiVector(map, rhs, lda, num_vecs); FORTRILINOS_CHECK_IERR()

  ! Step 0: create a handle
  solver_handle = TrilinosSolver(); FORTRILINOS_CHECK_IERR()

  ! ------------------------------------------------------------------
  ! Explicit setup and solve
  ! ------------------------------------------------------------------

  ! Step 1: initialize a handle
  call solver_handle%init(comm); FORTRILINOS_CHECK_IERR()

  ! Step 2: setup the problem
  call solver_handle%setup_matrix(A); FORTRILINOS_CHECK_IERR()

  ! Step 3: setup the solver
  call plist%set("Preconditioner Type", "None")
  call solver_handle%setup_solver(plist); FORTRILINOS_CHECK_IERR()

  ! Step 4: solve the system
  call X%randomize()
  call solver_handle%solve(B, X); FORTRILINOS_CHECK_IERR()

  ! Check the solution
  allocate(norms(1))
  call X%update(-one, Xtrue, one); FORTRILINOS_CHECK_IERR()
  call X%norm2(norms); FORTRILINOS_CHECK_IERR()

  ! TODO: Get the tolerance out of the parameter list
  if (norms(1) > 1e-6) then
    write(error_unit, '(A)') 'The solver did not converge to the specified residual!'
    stop 1
  end if

  ! Step 5: clean up
  call solver_handle%finalize(); FORTRILINOS_CHECK_IERR()

  ! ------------------------------------------------------------------
  ! Implicit (inversion-of-control) setup [ no solve ]
  ! ------------------------------------------------------------------
  allocate(op, source=TriDiagOperator(map, map))
  ! Step 1: initialize a handle
  call solver_handle%init(comm); FORTRILINOS_CHECK_IERR()

  ! Step 2: setup the problem
  ! Implicit (inversion-of-control) setup
  call init_ForTpetraOperator(op); FORTRILINOS_CHECK_IERR()
  call solver_handle%setup_operator(op); FORTRILINOS_CHECK_IERR()

  ! Step 3: setup the solver
  ! We cannot use most preconditioners without a matrix, so
  ! we remove any from the parameter list
  call plist%set("Preconditioner Type", "None")
  call solver_handle%setup_solver(plist); FORTRILINOS_CHECK_IERR()

  ! Step 4: solve the system
  call X%randomize()
  call solver_handle%solve(B, X); FORTRILINOS_CHECK_IERR()

  call X%update(-one, Xtrue, one); FORTRILINOS_CHECK_IERR()
  call X%norm2(norms); FORTRILINOS_CHECK_IERR()
  if (norms(1) > 1e-10) then
    write(error_unit, '(A)') 'The implicit result differs from explicit!'
    write(error_unit, '(A, ES14.7)') '   diff_norm =', norms(1)
    stop 666
  end if

  ! Step 5: clean up
  call solver_handle%finalize(); FORTRILINOS_CHECK_IERR()

  call op%release(); FORTRILINOS_CHECK_IERR()
  deallocate(op)
  ! ------------------------------------------------------------------

  call solver_handle%release(); FORTRILINOS_CHECK_IERR()
  call plist%release(); FORTRILINOS_CHECK_IERR()
  call X%release(); FORTRILINOS_CHECK_IERR()
  call B%release(); FORTRILINOS_CHECK_IERR()
  call A%release(); FORTRILINOS_CHECK_IERR()
  call map%release(); FORTRILINOS_CHECK_IERR()
  call comm%release(); FORTRILINOS_CHECK_IERR()
  deallocate(norms)
  deallocate(cols)
  deallocate(vals)
  deallocate(lhs)
  deallocate(rhs)

#ifdef HAVE_MPI
  ! Finalize MPI must be called after releasing all handles
  call MPI_FINALIZE(ierr)
#endif


end program
